desktopexporter/trace_store.go

```diff
@@ -23,6 +23,18 @@ func NewTraceStore(maxQueueSize int) *TraceStore {
 	}
 }
 
+func (store *TraceStore) AddMetric(_ context.Context, md MetricData) {
```

Comment:
Would it make sense to create a separate `MetricStore` and `LogStore`? I expect these are going to look significantly different given the different domains

Reply:
That would definitely be one option. I was trying to avoid creating separate data stores and considered adding a telemetry store instead. See https://github.com/CtrlSpice/otel-desktop-viewer/compare/main...codeboten:otel-desktop-viewer:codeboten/add-telemetry-data

Reply:
I suspect that wrapping things up in a [`TelemetryData` struct](https://github.com/CtrlSpice/otel-desktop-viewer/compare/main...codeboten:otel-desktop-viewer:codeboten/add-telemetry-data#diff-e4565670e1a3c1ca4d10c639e0dce343b7558f352ee509a40b2abaa73cc84c3aR15) is going to lead to some more complicated code paths since you will need to either switch on the type or assert on the type frequently. (I might feel differently if go had support for Algebraic Data Types)

I also suspect that we'll want to treat them differently. Eg. for metrics we would probably want an index of all metrics names we've received, and the ability to ask for metrics payloads filtered by name.

Is there some benefit to putting them together in one structure that I'm missing? When would we not be looking for a specific kind of data? eg: recent traces, recent logs, list of all metrics received.

Reply:
For the purpose of the desktop viewer, I was thinking of the telemetry as a single stream of data coming into the UI, where I wouldn't necessarily want to differentiate upfront what they are. The UI would just show a list (on the left) of all the recent data points received, regardless of their types. The information contained in the details (right side) and the center panel would change depending on the type of the data.

> is going to lead to some more complicated code paths since you will need to either switch on the type or assert on the type frequently

I wonder if this is something generics could help with simplifying.

Reply:
> For the purpose of the desktop viewer, I was thinking of the telemetry as a single stream of data coming into the UI, where I wouldn't necessarily want to differentiate upfront what they are.

That's an interesting idea 🤔 @CtrlSpice and I were thinking of breaking things down by signal: Traces, Metrics, Logs. Breaking things down by payload feels a bit like the protocol leaking into and defining the UI instead of starting from the end-users mental model. I could see a list of everything being helpful to someone working on the collector, but it feels like it might be overwhelming for an OTel end-user working on their application? Offering one doesn't prohibit also offering the other though.

In my mental model, showing recently received payloads maps fairly well to traces, as the UI currently works. Traces that have recently received spans bubble to the top, and completed traces age out.

For metrics I would expect a payload periodically, but the UI would show a list of all the names of metrics ever received. Clicking on one would visualize the metric over time and from there dig into the payload values if you wanted details (bonus points for visualizations of payloads like histograms), or allow you to filter on attributes / resource attributes.

I haven't dug into the OTel log data model yet, but breaking a stream of logs down into individual payloads in the UI feels odd. I am heavily influenced by the way Heroku treats and renders logs here however.

Of course, it would be nice for this tool to be useful to people working at the protocol level too! Perhaps a configuration flag could be used to enable / disable certain views for different audiences? The primary target audience for the CLI has been new developers.

I don't think anything in your data model would prevent the APIs driving the views I'm imagining from being built, so I'm okay going in this direction. Thank you for sharing your reasoning here.